staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -36,53 +33,57 @@ type Route struct {
 	// TODO: add more fields here if needed
 }
 
-func getRoutes(input io.Reader) ([]Route, error) {
-	routes := []Route{}
-	if input == nil {
-		return nil, fmt.Errorf("input is nil")
-	}
-	scanner := bufio.NewReader(input)
-	for {
-		line, err := scanner.ReadString('\n')
-		if err == io.EOF {
-			break
-		}
-		//ignore the headers in the route info
-		if strings.HasPrefix(line, "Iface") {
-			continue
-		}
-		fields := strings.Fields(line)
-		routes = append(routes, Route{})
-		route := &routes[len(routes)-1]
-		route.Interface = fields[0]
-		ip, err := parseIP(fields[1])
-		if err != nil {
-			return nil, err
-		}
-		route.Destination = ip
-		ip, err = parseIP(fields[2])
-		if err != nil {
-			return nil, err
-		}
-		route.Gateway = ip
+const (
+	ipCommand   = "/usr/sbin/ip"
+	grepCommand = "/usr/bin/grep"
+)
+
+var (
+	ipv4_zero = net.IP{0, 0, 0, 0}
+	ipv6_zero = net.ParseIP("::")
+)
+
+var execCommand = exec.Command
+
+// buildRouteCommand forms IP route command(s) to obtain routes and filter
+// on the default routes. Can do IPv4 only, or IPv4 + IPv6.
+func buildRouteCommand(withIPv6 bool) string {
```

Comment:
I would say yes to the first question (backwards compatibility), and a probably not for the second (since this is dual stack we're considering... IPv6-only being a separate consideration).
I think there are 2 ways to do the precedence, depending on whether backwards compatibility is more important, or valuing a configured default route (for IPv6) is more important. What I had in mind with my earlier comments was this order:
- IPv4 addr with default route
- IPv6 addr with default route (trust the intent, since an explicit def route is config'd)
- IPv4 addr (w/o def route) (Fallback choice)
- IPv6 addr (w/o def route) (Fallback choice)
The other option would be:
- IPv4 addr with default route
- IPv4 addr (w/o def route) (Fallback choice)
- IPv6 addr with default route
- IPv6 addr (w/o def route) (Fallback choice)
The second option is safer for backwards compatibility, but forces the operator into IPv6-only if they want IPv6 node addresses to be chosen.